pkg/causality/worker.go

```diff
@@ -14,11 +14,15 @@
 package causality
 
 type (
-	conflictKey = int64
+	conflictKey = uint64
 )
 
 type txnEvent interface {
-	ConflictKeys() []conflictKey
+	// Keys are in range [0, numSlots) and must be deduped.
+	//
+	// NOTE: if the conflict detector is accessed by multiple threads concurrently,
+	// ConflictKeys must also be sorted.
+	ConflictKeys(numSlots uint64) []conflictKey
```

Comment:
Why do they have to be sorted?

Reply:
I removed the global lock on `slots`. Now slots of a transaction are locked one by one. If the order can't be guaranteed, it's possible that A locks `slot_1` and requires `slot_2` but B locks `slot_2` and requires `slot_1`.